backend-es/test/snapshots-out/Snapshot.FunctionCompose03.js

```diff
@@ -1,23 +1,23 @@
 const test4 = f => g => {
   const $0 = g();
   const $1 = f();
-  const $2 = g();
-  const $3 = f();
+  const $2 = f();
+  const $3 = g();
```

Comment:
I still find this odd that these definitions are being swapped around. I'm not sure if that's incorrect, it just seems like a strange artifact. Do you have any idea what is happening here? It may be easier to tell if all the invocations were different functions.

Reply:
Just making sure things are correct & then I'll respond to the point about why this is happening:

In the old one, 0 == g, 1 == f, 2 == g, 3 == f, 4 == g, so the final result of 0,1,2,3,4 was g,f,g,f,g.
In the current one, 0 == g, 1 == f, 2 == f, 3 == g, 4 == f, so the final result of 0,2,3,1,4 is g, f, g, f, g.

So it's correct, but as to why it's happening, I don't know.

Here's the trace from the new one:

```bash
== Snapshot.FunctionCompose03 ==================================================
================================================================================

++ Snapshot.FunctionCompose03.test4 Step 1 (Original) ++++++++++++++++++++++++++
  \f@0 g@1 ->
    Snapshot.FunctionCompose03.compose
      ( Snapshot.FunctionCompose03.compose
        ( Snapshot.FunctionCompose03.compose
          (g@1 Data.Unit.unit)
          (f@0 Data.Unit.unit)
        )
        (g@1 Data.Unit.unit)
      )
      ( Snapshot.FunctionCompose03.compose
        (f@0 Data.Unit.unit)
        (g@1 Data.Unit.unit)
      )

-- Snapshot.FunctionCompose03.test4 Step 2 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = f@0 #[prim.undefined] in
    {- Inline -}
    let @5 = \x@5 -> @2 (@4 x@5) in
    {- Inline -}
    let
      @6 =
        let
          @6 =
            {- Inline -}
            let
              @6 =
                let @6 = g@1 #[prim.undefined] in
                \x@7 -> @5 (@6 x@7) in
            @6 in
        let @7 = g@1 #[prim.undefined] in
        {- Inline -}
        let @8 = \x@8 -> @3 (@7 x@8) in
        \x@9 -> @6 (@8 x@9) in
    @6

-- Snapshot.FunctionCompose03.test4 Step 3 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = f@0 #[prim.undefined] in
    let @5 = g@1 #[prim.undefined] in
    {- Inline -}
    let
      @6 =
        \x@6 ->
          {- Inline -}
          let @7 = @5 x@6 in
          @2 (@4 @7) in
    let @7 = g@1 #[prim.undefined] in
    \x@8 -> @6 (@3 (@7 x@8))

-- Snapshot.FunctionCompose03.test4 Step 4 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = f@0 #[prim.undefined] in
    let @5 = g@1 #[prim.undefined] in
    let @6 = g@1 #[prim.undefined] in
    \x@7 ->
      {- Inline -}
      let @8 = @3 (@6 x@7) in
      @2 (@4 (@5 @8))

-- Snapshot.FunctionCompose03.test4 Step 5 (Final) -----------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = f@0 #[prim.undefined] in
    let @5 = g@1 #[prim.undefined] in
    let @6 = g@1 #[prim.undefined] in
    \x@7 -> @2 (@4 (@5 (@3 (@6 x@7))))
```

and here's the old one:

```bash
== Snapshot.FunctionCompose03 ==================================================
================================================================================

++ Snapshot.FunctionCompose03.test4 Step 1 (Original) ++++++++++++++++++++++++++
  \f@0 g@1 ->
    Snapshot.FunctionCompose03.compose
      ( Snapshot.FunctionCompose03.compose
        ( Snapshot.FunctionCompose03.compose
          (g@1 Data.Unit.unit)
          (f@0 Data.Unit.unit)
        )
        (g@1 Data.Unit.unit)
      )
      ( Snapshot.FunctionCompose03.compose
        (f@0 Data.Unit.unit)
        (g@1 Data.Unit.unit)
      )

-- Snapshot.FunctionCompose03.test4 Step 2 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    {- Inline -}
    let @4 = \x@4 -> @2 (@3 x@4) in
    let
      @5 =
        {- Inline -}
        let
          @5 =
            let @5 = g@1 #[prim.undefined] in
            \x@6 -> @4 (@5 x@6) in
        @5 in
    {- Inline -}
    let
      @6 =
        let @6 = f@0 #[prim.undefined] in
        let
          @7 =
            let @7 = g@1 #[prim.undefined] in
            {- Inline -}
            let @8 = \x@8 -> @6 (@7 x@8) in
            @8 in
        \x@8 -> @5 (@7 x@8) in
    @6

-- Snapshot.FunctionCompose03.test4 Step 3 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = g@1 #[prim.undefined] in
    {- Inline -}
    let
      @5 =
        \x@5 ->
          {- Inline -}
          let @6 = @4 x@5 in
          @2 (@3 @6) in
    let @6 = f@0 #[prim.undefined] in
    let @7 = g@1 #[prim.undefined] in
    {- Inline -}
    let @8 = \x@8 -> @6 (@7 x@8) in
    \x@9 -> @5 (@8 x@9)

-- Snapshot.FunctionCompose03.test4 Step 4 -------------------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = g@1 #[prim.undefined] in
    let @5 = f@0 #[prim.undefined] in
    let @6 = g@1 #[prim.undefined] in
    \x@7 ->
      {- Inline -}
      let @8 = @5 (@6 x@7) in
      @2 (@3 (@4 @8))

-- Snapshot.FunctionCompose03.test4 Step 5 (Final) -----------------------------
  \f@0 g@1 ->
    let @2 = g@1 #[prim.undefined] in
    let @3 = f@0 #[prim.undefined] in
    let @4 = g@1 #[prim.undefined] in
    let @5 = f@0 #[prim.undefined] in
    let @6 = g@1 #[prim.undefined] in
    \x@7 -> @2 (@3 (@4 (@5 (@6 x@7))))
```

Just reading through those, I can't figure out exactly what's going on yet.

Reply:
It's odd to me that in Step 2 there are a bunch of left-associative binds (level 6). I wouldn't expect that to happen since `eval` should be reassociating all binds.